models/User.go

```diff
@@ -1414,69 +1415,151 @@ func PerformMortgageTransaction(userId, stockId uint32, stockQuantity int64) (*T
+		sql = "Select sum(stocksInBank) as sc from MortgageDetails where UserId=? and StockId=?"
+		err = db.Raw(sql, user.Id, stockId).Scan(&mortgageCount).Error
 		if err != nil {
 			l.Error(err)
 			return nil, err
 		}
 
-		if stockOwned < -stockQuantity {
-			l.Errorf("Insufficient stocks in ownership. Have %d, want %d", stockOwned, stockQuantity)
+		if AbsBranch(stockQuantity) > stockCount.Sc-mortgageCount.Sc {
+			l.Errorf("Insufficient stocks to mortgage. Have %d, want %d", stockCount.Sc, stockQuantity)
 			return nil, NotEnoughStocksError{}
 		}
 	}
 
-	trTotal := -int64(mortgagePrice) * stockQuantity * int64(rate) / 100
-	l.Infof("lalalalalalalala %d %d %d", trTotal, stockQuantity, int64(rate/100))
-	if int64(user.Cash)+trTotal < 0 {
-		l.Debugf("User does not have enough cash. Want %d, Have %d. Failing.", trTotal, user.Cash)
-		return nil, NotEnoughCashError{}
+	var trTotal int64
+
+	/* It is retrive action; This block will set correct trTotal and stockQuantity equal
+	to amount of stocks that could be retrieved; This block also modifies MortgageDetails
+	table, btw this table is not used anywhere except in current function */
+
+	if stockQuantity >= 0 {
+
+		db := getDB()
+		sql := "SELECT id, stocksInBank, mortgagePrice from MortgageDetails where userId=? AND stockId=? ORDER BY mortgagePrice"
+		rows, err := db.Raw(sql, user.Id, stockId).Rows()
+		if err != nil {
+			l.Error(err)
+			return nil, err
+		}
+		defer rows.Close()
+
+		// price represents cost of stock at time of mortgage
+		// stocksInBank represents number of stocks mortgaged at that time
+		var id int32
+		var price int64
+		var stocksInBank int64
+
+		// tempStockQuantity is stock quantity left to retrieve
+		tempStockQuantity := stockQuantity
+		tempUserCash := int64(user.Cash)
+		l.Infof("Cash %d", tempUserCash)
+		var maxStocksRetrieval int64
+
+		for rows.Next() {
+			rows.Scan(&id, &stocksInBank, &price)
+			maxStocksRetrieval = tempUserCash / price
+
+			if maxStocksRetrieval == 0 || tempStockQuantity == 0 {
+				stockQuantity -= tempStockQuantity // Amount of stocks retrieved till now
+				break
+			}
+
+			if maxStocksRetrieval > stocksInBank {
+				maxStocksRetrieval = stocksInBank
+			}
+
+			if maxStocksRetrieval > tempStockQuantity { // We don't want to retrieve more than required
+				maxStocksRetrieval = tempStockQuantity
+			}
+			expense := -int64(price) * int64(maxStocksRetrieval) * MORTGAGE_RETRIEVE_RATE / 100
+			trTotal += expense
+			if maxStocksRetrieval < stocksInBank {
+				sql := "UPDATE MortgageDetails SET stocksInBank=? where id=?"
+				err = db.Exec(sql, stocksInBank-maxStocksRetrieval, id).Error
+				if err != nil {
+					l.Infof("PerformMortgageTransaction failed for userId = %d, stockId = %d amount = %d while updating retrieved stocks", userId, stockId, stockQuantity)
+					l.Error(err)
+					return nil, err
+				}
+			} else /* So we can delete that entire row */ {
+				sql := "DELETE from MortgageDetails WHERE id=?"
```

Comment:
If you delete from here, and you don't update the Transactions table, the invariant I mentioned earlier gets violated. Secondly, we don't have any trail to follow later if we need to understand what happened.

Reply:
Using database transaction now